File: packages/faasit-runtime/javascript/src/Workflow.ts
import { FaasitRuntime } from "./runtime/FaasitRuntime"
import { FunctionContainerConfig } from "./type"

type HandlerType = (frt: FaasitRuntime) => unknown

type DeclarativeWorkflowBuilder = () => unknown

export interface WorkflowBuilder {
  // configure function
  func(name: string): WorkflowFuncBuilder

  // get all valid function names
  get_funcs(): string[]

  // configure executor
  executor(): WorkflowExecutorBuilder

  // finalize and output workflow spec (used by engine)
  build(): WorkflowSpec
}

export interface WorkflowFuncBuilder {
  // add handler
  set_handler(fn: HandlerType): WorkflowFuncBuilder
}

export interface WorkflowExecutorBuilder {
  // use custom executor to replace auto generated executor
  set_custom_handler(fn: HandlerType): WorkflowExecutorBuilder

  // use declarative workflow description language to describe a workflow
  // TODO: design declarative and higher level api to describe workflow
  set_declarative(): WorkflowExecutorBuilder
}

export interface WorkflowFunc {
  name: string
  handler: HandlerType
}

export interface WorkflowExecutor {
  kind: 'e_custom'
  handler: HandlerType
}

// TODO: design details for WorkflowSpec
export interface WorkflowSpec {
  functions: WorkflowFunc[]
  exeuctor: WorkflowExecutor
}

class SimpleWorkflowBuilder implements WorkflowBuilder {
  private spec: WorkflowSpec = {
    functions: [],
    exeuctor: {
      kind: 'e_custom',
      handler: () => { throw new Error(`not implemented executor`) }
    }
  }
  constructor() { }

  func(name: string): WorkflowFuncBuilder {
    if (name.startsWith('__')) {
      throw new Error(`function name can not start with __, name=${name}`)
    }

    const func: Partial<WorkflowFunc> = {
      name
    }

    this.spec.functions.push(func as WorkflowFunc)

    return {
      set_handler(fn) {
        func.handler = fn;
        return this
      },
    }
  }
  get_funcs(): string[] {
    return this.spec.functions.map(v => v.name)
  }
  executor(): WorkflowExecutorBuilder {
    const spec = this.spec
    return {
      set_custom_handler(fn) {
        spec.exeuctor.handler = fn
        return this
      },
      set_declarative() {
        throw new Error("Method not implemented.")
      },
    }
  }
  build(): WorkflowSpec {
    return this.spec
  }
}

type WorkflowBuilderType = (builder: WorkflowBuilder) => WorkflowSpec
export function createWorkflow(fn: WorkflowBuilderType): WorkflowSpec {
  const builder = new SimpleWorkflowBuilder()
  return fn(builder)
}

// Run workflow function in each container
export class WorkflowContainerRunner {
  constructor(private containerConf: FunctionContainerConfig, private spec: WorkflowSpec) {
  }

  run(frt: FaasitRuntime): unknown {
    const { funcName } = this.containerConf.workflow
    const func = this.route(funcName)
    return func(frt)
  }

  route(name: string): HandlerType {
    // run executor
    if (name === '__executor') {
      return this.spec.exeuctor.handler
    }

    const target = this.spec.functions.find(v => v.name === name)

    if (!target) {
      throw new Error(`no such workflow target, name = ${name}`)
    }

    return target.handler
  }
}

File: packages/faasit-runtime/python/faasit_runtime/workflow/workflow.py
from typing import Callable
from faasit_runtime.runtime.faasit_runtime import FaasitRuntime, FaasitResult
from faasit_runtime.utils import config



class WorkFlowFunc:
    def __init__(self, name: str, handler: Callable[[FaasitRuntime], FaasitResult] = None) -> None:
        self.name = name
        self.handler = handler

    def set_handler(self, 
                    handler: Callable[[FaasitRuntime], FaasitResult]):
        self.handler = handler

class WorkFlowExecutor(WorkFlowFunc):
    def __init__(self, name: str, handler: Callable[[FaasitRuntime], FaasitResult] = None) -> None:
        super().__init__(name, handler)

    def set_handler(self, 
                    handler: Callable[[FaasitRuntime], FaasitResult]):
        super().set_handler(handler)

class WorkFlow:
    def __init__(self, 
                 functions: list[WorkFlowFunc],
                 executor: WorkFlowExecutor) -> None:
        self.functions = functions
        self.executor = executor
        pass

class WorkFlowBuilder:
    def __init__(self) -> None:
        self.funcs: list[WorkFlowFunc] = []
        self.executor_func: WorkFlowExecutor = None
        pass

    # This method is used to add a function to the workflow
    def func(self, funcName:str) -> WorkFlowFunc:
        # create a new function
        newFunc = WorkFlowFunc(funcName)
        self.funcs.append(newFunc)
        return newFunc

    # get all the funcs in the workflow
    def get_funcs(self) -> list[WorkFlowFunc]:
        return self.funcs()

    # add an executor to the workflow
    def executor(self) -> WorkFlowExecutor:
        exe_func = WorkFlowFunc('__executor')
        # self.funcs.append(exe_func)
        self.executor_func = exe_func
        return exe_func

    # build the workflow
    def build(self) -> WorkFlow:
        if self.executor_func == None:
            raise ValueError('Executor function not set')
        return WorkFlow(self.funcs, self.executor_func)
    
class WorkFlowRunner:
    def __init__(self, workflow: WorkFlow) -> None:
        self.conf = config.get_function_container_config()
        self.workflow = workflow
        pass

    def run(self, frt: FaasitRuntime, *args) -> FaasitResult:
        funcName = self.conf['funcName']
        fn = self.route(funcName)
        return fn(frt, *args)
    
    def route(self, name: str) -> Callable[[FaasitRuntime], FaasitResult]:
        if name == '__executor':
            return self.workflow.executor.handler
        for func in self.workflow.functions:
            if func.name == name:
                return func.handler
        raise ValueError(f'Function {name} not found in workflow')